src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/SQLEngineAvroToStructuredTransformer.java

```diff
@@ -0,0 +1,103 @@
+ * the License.
+ */
+
+package io.cdap.plugin.gcp.bigquery.sqlengine;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.etl.api.engine.sql.SQLEngineException;
+import io.cdap.plugin.gcp.bigquery.source.BigQueryAvroToStructuredTransformer;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import javax.annotation.Nullable;
+
+/**
+ * Create StructuredRecords from GenericRecords when pulling records from BigQuery.
+ *
+ * The following transformations are supported:
+ * BigQuery INT64 -> Integer (with Overflow check)
+ * BigQuery FLOAT64 -> Float (with Overflow check)
+ *
+ * Otherwise, field conversion gets delegated to the {@link BigQueryAvroToStructuredTransformer}
+ */
+public class SQLEngineAvroToStructuredTransformer extends BigQueryAvroToStructuredTransformer
+  implements Serializable {
+
+  private static final String LONG_TO_INTEGER_CONVERSION_ERROR =
+    "Could not convert Long value '%d' into Integer when reading from BigQuery";
+  private static final String INTEGER_CONVERSION_ERROR =
+    "Could not convert Object '%s' into Integer when reading from BigQuery";
+  private static final String DOUBLE_TO_FLOAT_CONVERSION_ERROR =
+    "Could not convert Double '%f' into Float when reading from BigQuery";
+  private static final String FLOAT_CONVERSION_ERROR =
+    "Could not convert Object '%s' into Float when reading from BigQuery";
+
+  @Override
+  @Nullable
+  protected Object convertField(Object field, Schema fieldSchema) throws IOException {
+    if (field == null) {
+      return null;
+    }
+
+    // Handle Int types
+    if (fieldSchema.getType() == Schema.Type.INT) {
+      return mapInteger(field);
+    }
+
+    // Handle float types
+    if (fieldSchema.getType() == Schema.Type.FLOAT) {
+      return mapFloat(field);
+    }
+
+    // Handle Strings that are stored as a Byte Buffer.
+    if (fieldSchema.getType() == Schema.Type.STRING && field instanceof ByteBuffer) {
+      return StandardCharsets.UTF_8.decode((ByteBuffer) field).toString();
+    }
+
+    return super.convertField(field, fieldSchema);
+  }
+
+  @VisibleForTesting
+  protected Integer mapInteger(Object field) throws SQLEngineException {
+    if (field instanceof Long) {
+      Long value = (Long) field;
+      Integer result = value.intValue();
+      if ((long) result != value) {
+        throw new SQLEngineException(String.format(LONG_TO_INTEGER_CONVERSION_ERROR, value));
+      }
+      return result;
+    } else {
+      throw new SQLEngineException(String.format(INTEGER_CONVERSION_ERROR, field));
+    }
+  }
+
+  @VisibleForTesting
+  protected Float mapFloat(Object field) {
+    if (field instanceof Double) {
+      Double value = (Double) field;
+      Float result = value.floatValue();
+      if ((double) result != value) {
```

Comment:
```groovy
double d = 12345/10000.0
float f = (float)d
double d2 = f
if (d2 != d) {
    println d + " != " + f
}
println Double.doubleToLongBits(d)
println Double.doubleToLongBits(d2)
```
Output:
```
1.2345 != 1.2345
4608238512912635789
4608238513140269056
```

Reply:
Note: this is groovy code, you can run it in Tools -> Groovy Console in Intellij.

Reply:
Hmm, good point. I had skipped over these float/double arithmetic quirks.

Do you have any suggestions on how to validate? The other strategy I see is to make sure the double falls "in range" with a `Float.MIN_VALUE >= double && Float.MAX_VALUE <= double`

Reply:
Could you clarify what are you trying to validate here? For floating point there is also loss of precision when you convert it to float. 
FYI: Float.MIN_VALUE is 2^-149 (positive)

Reply:
BigQuery stores all floats as Float64, This means Float fields in CDAP get stored as Float64 in BQ.

Since records in BQ are stored as Float64, they get mapped to Double when read from Avro.

My goal here is to ensure that this Double "fits" into a Float.

Reply:
In what cases it won't fit and what should happen? Where does this joiner output schema come from? Can user set joiner schema output be float with double as input and what would happen for BQ pushdown enabled vs disabled?

Reply:
I don't expect this to be an issue with joins. However, in the future, when aggregations are supported, we need to account for the posibility of something like an "overflow" (where, for example `SUM(some_int_column)` results in an integer overflow).

I changed the logic to ensure that the Float is "close enough" to the double value (within 1e-15) in order to assume the conversion from double to float is "valid".